tracer/test/Datadog.Trace.Tests/Telemetry/TelemetryDataBuilderTests.cs

```diff
@@ -0,0 +1,144 @@
+﻿// <copyright file="TelemetryDataBuilderTests.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Datadog.Trace.Telemetry;
+using FluentAssertions;
+using FluentAssertions.Execution;
+using Xunit;
+
+namespace Datadog.Trace.Tests.Telemetry
+{
+    public class TelemetryDataBuilderTests
+    {
+        private readonly ApplicationTelemetryData _application;
+        private readonly HostTelemetryData _host;
+
+        public TelemetryDataBuilderTests()
+        {
+            _application = new ApplicationTelemetryData { Env = "integration-ci", ServiceName = "Test Service" };
+            _host = new HostTelemetryData();
+        }
+
+        [Fact]
+        public void WhenNoApplicationData_Throws()
```

Comment:
Should `builder.BuildAppClosingTelemetryData` also throw `ArgumentNullException` when the application argument is null?

Reply:
I suppose we would never get to the point where we call `builder.BuildAppClosingTelemetryData` with a null application argument because we would first throw when calling `builder.BuildTelemetryData`. Is that the reasoning here?

Reply:
> Is that the reasoning here

I wish, this is discrepancy is a result of many refactors 😆 

Looking at this again, I think this is either:

- Unnecessary defensiveness, and we can remove the throwing here, as it logically can't be null anyway in the control flow + it's only used for serialization
- Not enough defensiveness, and we should throw for both functions, and we should also throw if host is missing (as that's also required)

I've done the latter in https://github.com/DataDog/dd-trace-dotnet/pull/2241/commits/092dc5a862b4886cbea782569d100c9744775910, but I'd be easily convinced to take the former instead 😉

Reply:
Either approach seems fine to me. I'm okay with taking a stronger defensive stance by default so the changes you made look good 😄